rfcs/2021-06-29-7709-helm-update-vector-config-pattern.md

```diff
@@ -0,0 +1,174 @@
+          filesystems:
+            excludes: ["binfmt_misc"]
+          mountpoints:
+            excludes: ["*/proc/sys/fs/binfmt_misc"]
+      internal_metrics:
+        type: internal_metrics
+      kubernetes_logs:
+        type: kubernetes_logs
+    sinks:
+      prometheus_sink:
+        type: prometheus_exporter
+        inputs: ["host_metrics", "internal_metrics"]
+        address: 0.0.0.0:9090
+  {{- end }}
+{{- end }}
+```
+
+To make it easier for users to make minor changes to our default configuration we could provide a copy of the
+default configuration (commented out). Users would uncomment the block and be able to make any required changes
+without having to rewrite the entire configuration.
+
+```yaml
+# Specify custom contents for the Vector config
+## ref: https://vector.dev/docs/reference/configuration/
+## Note a complete and valid configuration is required
+customConfig: {}
+  #  data_dir: "/vector-data-dir"
+  #  log_schema:
+  #    host_key: host
+  #    message_key: message
+  #    source_type_key: source_type
+  #    timestamp_key: timestamp
+  #  sources:
+  #    host_metrics:
+  #      type: host_metrics
+  #      filesystem:
+  #        devices:
+  #          excludes: ["binfmt_misc"]
+  #        filesystems:
+  #          excludes: ["binfmt_misc"]
+  #        mountpoints:
+  #          excludes: ["*/proc/sys/fs/binfmt_misc"]
+  #    internal_metrics:
+  #      type: internal_metrics
+  #    kubernetes_logs:
+  #      type: kubernetes_logs
+  #  sinks:
+  #    prometheus_sink:
+  #      type: prometheus_exporter
+  #      inputs: ["host_metrics", "internal_metrics"]
+  #      address: 0.0.0.0:9090
+```
+
+We will be able to support the existing deprecated fields until the 1.0 release by not rendering the new YAML based
+configuration if any deprecated config key is used or enabled, and if a combination of new and old keys are used we
+can use a `fail` function to terminate chart rendering early.
+
+## Doc-level Proposal
+
+The only page we currently have with documentation around Helm is on the Kubernetes platform installation page, which
+today does not mention how to configure any of the default components.
+
+## Rationale
+
+- Provides the easiest mechanism for users to bring their own configuration and disable our provided default config
+- The entire vector configuration will be in a single location for better readability and debuggability
+  - We remove a source of bugs by removing any merging of configuration options (between default provided and user provided)
+- Reduces the templating logic for the entire configuration file to an `if/else` statement and a one-liner to template the provided configuration (if needed)
+- By passing the `customConfig` through a `tpl` function users can more easily generate parts of their configuration from other values (ports, volumes, etc)
+- Matches the [Datadog Agent configuration](https://github.com/DataDog/helm-charts/blob/master/charts/datadog/values.yaml#L1023-L1048)
```

Comment:
There is also small amount of logic in the config map template https://github.com/DataDog/helm-charts/blob/master/charts/datadog/templates/datadog-yaml-configmap.yaml but overall I think that a good strategy for flexible configuration.

Reply:
@prognant were you talking about any specific templating in the configmap? We may include some to ensure ports match configuration and such but I don't think we'll need much.